src/bios.rs

```diff
@@ -142,10 +144,63 @@ impl Component for Bios {
         crate::component::query_adopt_state()
     }
 
+    // Backup the current grub.cfg and install new static config
+    // - Backup "/boot/loader/grub.cfg" to "/boot/grub2/grub.cfg.bak"
+    // - Remove symlink "/boot/grub2/grub.cfg"
+    // - Install new static grub.cfg
+    // - Remove old "/boot/loader/grub.cfg"
+    fn migrate_static_grub_config(&self, sysroot_path: &str, destdir: &openat::Dir) -> Result<()> {
+        use anyhow::Context;
+        use openat_ext::OpenatDirExt;
+
+        // sysroot_path is /, destdir is Dir of /
+        let grub_config_path = Utf8PathBuf::from(sysroot_path).join("boot/grub2");
+        let grub_config_dir = destdir
+            .sub_dir("boot/grub2")
+            .context("Opening boot/grub2")?;
+
+        let grub_config = grub_config_path.join(grubconfigs::GRUBCONFIG);
+
+        // On BIOS, /boot/grub2/grub.cfg is symlink to /boot/loader/grub.cfg,
+        // backup it to /boot/grub2/grub.cfg.bak
+        if !grub_config.exists() {
+            anyhow::bail!("Not found '{}'", grub_config);
+        } else if !grub_config.is_symlink() {
+            anyhow::bail!("'{}' is not a symlink", grub_config);
+        }
+
+        // Only migrate if /boot/grub2/grub.cfg is symlink
+        let real_config = grub_config_dir.read_link(grubconfigs::GRUBCONFIG)?;
+        let real_config =
+            Utf8PathBuf::from_path_buf(real_config).expect("Path should be valid UTF-8");
+        // Resolve symlink location
+        let mut current_config = grub_config_path.clone();
+        current_config.push(real_config);
+        let backup_config = grub_config_path.join(grubconfigs::GRUBCONFIG_BACKUP);
+
+        // Backup the current GRUB config which is hopefully working right now
+        println!(
+            "Creating a backup of the current GRUB config '{}' in '{}'...",
+            current_config, backup_config
+        );
+        std::fs::copy(&current_config, &backup_config).context("Failed to backup GRUB config")?;
+
+        crate::grubconfigs::install(&destdir, None, true)?;
```

Comment:
Is this doing an atomic swap with the symlink?

Reply:
https://docs.rs/openat-ext/latest/openat_ext/trait.OpenatDirExt.html#method.write_file_contents
it should